### ISSUE TITLE
"Aborted (core dumped)" but works fine in interactive mode

### ISSUE THREAD

**Issue by jshanna100** [psychopy/psychopy#2344]
We recently upgraded to Psychopy 3.0.5 from 1.90.3, and our experiments - which had worked perfectly with 1.90.3 - only work when they are run in interactive mode (parameter -i with python at command line). 

Otherwise they crash almost immediately only with the message "Aborted (core dumped)"

**Comment by peircej** [comment]
Really hard to know what's going on from this information, but this isn't the case for most people (3.0.5 has been pretty robust for most experiments) so I think you'll need to tell us more about what this study is and what might be different from other common experiments. Does it have custom code? Does it use hardware? i.e., assuming that the basic demos work on your machine, what differs between your not-working experiment and the working demos?

**Comment by jshanna100** [comment]
Rather than installing from Conda as before, I manually installed from download, so now version 3.0.6. It still fails, but it at least gives a more specific message:

pygame 1.9.4
Hello from the pygame community. https://www.pygame.org/contribute.html
Fatal Python error: (pygame parachute) Segmentation Fault

Current thread 0x00007f0e8fd02740 (most recent call first):
  File "<frozen importlib._bootstrap>", line 222 in _call_with_frames_removed
  File "<frozen importlib._bootstrap_external>", line 938 in create_module
  File "<frozen importlib._bootstrap>", line 577 in module_from_spec
  File "<frozen importlib._bootstrap>", line 666 in _load_unlocked
  File "<frozen importlib._bootstrap>", line 957 in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 968 in _find_and_load
  File "/home/stim/anaconda3/envs/psychopy3/lib/python3.5/site-packages/PsychoPy-3.0.6-py3.5.egg/psychopy/sound/backend_sounddevice.py", line 13 in <module>
  File "<frozen importlib._bootstrap>", line 222 in _call_with_frames_removed
  File "<frozen importlib._bootstrap_external>", line 697 in exec_module
  File "<frozen importlib._bootstrap>", line 673 in _load_unlocked
  File "<frozen importlib._bootstrap>", line 957 in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 968 in _find_and_load
  File "<frozen importlib._bootstrap>", line 222 in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1015 in _handle_fromlist
  File "/home/stim/anaconda3/envs/psychopy3/lib/python3.5/site-packages/PsychoPy-3.0.6-py3.5.egg/psychopy/sound/__init__.py", line 79 in <module>
  File "<frozen importlib._bootstrap>", line 222 in _call_with_frames_removed
  File "<frozen importlib._bootstrap_external>", line 697 in exec_module
  File "<frozen importlib._bootstrap>", line 673 in _load_unlocked
  File "<frozen importlib._bootstrap>", line 957 in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 968 in _find_and_load
  File "<frozen importlib._bootstrap>", line 222 in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1015 in _handle_fromlist
  File "aud_attention.py", line 4 in <module>
Aborted (core dumped)

This is custom code, but it's failing already at the fourth line, during the imports. Up to that point the code is:

import numpy as np
from psychopy import visual, prefs, event, core, parallel
prefs.general["audioLib"] = ["pygame"]
from psychopy import sound

What is it about running python with or without -i which could be causing it to succeed/fail?
For info, this is running on Lubuntu 18.04, in the anaconda environment set up according to the website recommendations, with the exception that wxpython and psychopy are manually installed.

**Comment by jshanna100** [comment]
And the demos seem to run fine, btw.

**Comment by peircej** [comment]
OK, so it's clearly an issue loading audio. My guess is that the pygame audio lib isn't working properly on your installation (possibly on conda at all). Try using pyo or sounddevice sound libs instead. Not sure why the pygame sound was working before and not now, but my guess is that it got updated to a broken version during the install process.

**Comment by jshanna100** [comment]
Changing to pyo I get exactly the same issue. Pygame is in fact working, but only if I either run python in interactive mode, or put the code in line by line. I can for example initialise and play a wav file by hand with no trouble, which does not suggest that pygame is actually broken.

**Comment by peircej** [comment]
I'm afraid I don't know what to suggest here, not having an Lubuntu box to test on. :-/
It is on my list to give the regular Ubuntu app some love though (currently waiting on a new box to play with) and maybe your issue will get fixed as I work on that.

**Comment by jshanna100** [comment]
I don't know why, but if you import sound before core and visual, then it works. Something about importing sound after pygame has already been imported (presumably visual uses pygame?) causes it to crash, so just import sound before that.

This still doesn't explain why it works with whatever import order when you run python with -i, but nevertheless it's a solution.

**Comment by stilley2** [comment]
Assuming this is the same issue I've been having, I think i tracked it down to the `import readline` in psychopy/sound/backend_sounddevice.py . For some reason when this import occurs matters. If I add `import readline` to the main `psychopy` launch script the error goes away

**Comment by peircej** [comment]
OK, this looks very much like an issue with sounddevice and readline, for which there's a lot more detail in these issues:

https://github.com/spatialaudio/python-sounddevice/issues/166
https://github.com/psychopy/psychopy/issues/2230
https://github.com/psychopy/psychopy/issues/2662

@hoechenberger in issue 2230 it turned out that importing readline in sound fixed the issue for some people but then broke for others (e.g. 2662). Maybe the solution is to import readline even earlier, like in `psychopy.__init__.py`?

**Comment by peircej** [comment]
OK. For 2020.1.1 release I'm going to try putting `import readline` right into the `psychopy.__init__` in the hope this will be fixed once and for all (even though it might confuse some people and code-linters seeing it there!). I'll close this issue for now but we can reopen if it doesn't fix it.

**Comment by hoechenberger** [comment]
Thanks @peircej, sorry I didn't get back to you, must've missed that message. Let's hope that change fixes the issue :)
